From 01564a47c0dae5b06a11217789459d7ebd1f61fa Mon Sep 17 00:00:00 2001
From: Jes Sorensen <Jes.Sorensen@redhat.com>
Date: Mon, 1 Sep 2014 09:25:15 -0400
Subject: [md] raid5: flush out all pending requests before proceeding with reshape

Message-id: <1409563584-9000-14-git-send-email-Jes.Sorensen@redhat.com>
Patchwork-id: 89950
O-Subject: [RHEL7.1 PATCH 13/82] md/raid5: flush out all pending requests before proceeding with reshape.
Bugzilla: 1085530
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

From: NeilBrown <neilb@suse.de>

Some requests - particularly 'discard' and 'read' are handled
differently depending on whether a reshape is active or not.

It is harmless to assume reshape is active if it isn't but wrong
to act as though reshape is not active when it is.

So when we start reshape - after making clear to all requests that
reshape has started - use mddev_suspend/mddev_resume to flush out all
requests.  This will ensure that no requests will be assuming the
absence of reshape once it really starts.

Signed-off-by: NeilBrown <neilb@suse.de>
(cherry picked from commit 4d77e3ba88d085836f1e8e475e3131844dd89d04)
Signed-off-by: Jes Sorensen <Jes.Sorensen@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/md/raid5.c b/drivers/md/raid5.c
index 1be4941..3b433d7 100644
--- a/drivers/md/raid5.c
+++ b/drivers/md/raid5.c
@@ -6287,6 +6287,13 @@ static int raid5_start_reshape(struct mddev *mddev)
  write_seqcount_end(&conf->gen_lock);
  spin_unlock_irq(&conf->device_lock);
 
+ /* Now make sure any requests that proceeded on the assumption
+  * the reshape wasn't running - like Discard or Read - have
+  * completed.
+  */
+ mddev_suspend(mddev);
+ mddev_resume(mddev);
+
  /* Add some new drives, as many as will fit.
   * We know there are enough to make the newly sized array work.
   * Don't add devices if we are reducing the number of
-- 
1.7.1